sdk/core/System.ClientModel/src/Message/PipelineRequestHeaders.cs

```diff
@@ -0,0 +1,140 @@
+
+    public override void Set(string name, string value)
+        => _headers.Set(new IgnoreCaseString(name), value);
+
+    public override void Add(string name, string value)
+    {
+        if (_headers.TryAdd(new IgnoreCaseString(name), value, out object? currentValue))
+        {
+            return;
+        }
+
+        switch (currentValue)
+        {
+            case string stringValue:
+                _headers.Set(new IgnoreCaseString(name), new List<string> { stringValue, value });
+                break;
+            case List<string> listValue:
+                listValue.Add(value);
+                break;
+        }
+    }
+
+    public override bool TryGetValue(string name, out string? value)
+    {
+        if (_headers.TryGetValue(new IgnoreCaseString(name), out object? headerValue))
+        {
+            value = GetHeaderValueString(name, headerValue);
+            return true;
+        }
+
+        value = default;
+        return false;
+    }
+
+    public override bool TryGetValues(string name, out IEnumerable<string>? values)
+    {
+        if (_headers.TryGetValue(new IgnoreCaseString(name), out object? value))
+        {
+            values = GetHeaderValueEnumerable(name, value);
+            return true;
+        }
+
+        values = default;
+        return false;
+    }
+
+    public override IEnumerator<KeyValuePair<string, string>> GetEnumerator()
+        => GetHeadersStringValues().GetEnumerator();
+
+    // Internal API provided to take advantage of performance-optimized implementation.
+    internal bool GetNextValue(int index, out string name, out object value)
+    {
+        if (index >= _headers.Count)
+        {
+            name = default!;
+            value = default!;
+            return false;
+        }
+
+        _headers.GetAt(index, out IgnoreCaseString headerName, out object headerValue);
+        name = headerName;
+        value = headerValue;
+        return true;
+    }
+
+    #region Implementation
+    private IEnumerable<KeyValuePair<string, string>> GetHeadersStringValues()
+    {
+        for (int i = 0; i < _headers.Count; i++)
+        {
+            _headers.GetAt(i, out IgnoreCaseString name, out object value);
+            string values = GetHeaderValueString(name, value);
+            yield return new KeyValuePair<string, string>(name, values);
+        }
+    }
+
+    private IEnumerable<KeyValuePair<string, IEnumerable<string>>> GetHeadersListValues()
+    {
+        for (int i = 0; i < _headers.Count; i++)
+        {
+            _headers.GetAt(i, out IgnoreCaseString name, out object value);
+            IEnumerable<string> values = GetHeaderValueEnumerable(name, value);
+            yield return new KeyValuePair<string, IEnumerable<string>>(name, values);
+        }
+    }
+
+    private static string GetHeaderValueString(string name, object value)
+        => value switch
+        {
+            string stringValue => stringValue,
```

Comment:
Similar comment regarding whether null values are permitted, and if so, we would need to handle it here.

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/41292